cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,544 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.securestore;
+
+import co.cask.cdap.api.security.securestore.SecureStore;
+import co.cask.cdap.api.security.securestore.SecureStoreData;
+import co.cask.cdap.api.security.securestore.SecureStoreManager;
+import co.cask.cdap.api.security.securestore.SecureStoreMetadata;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import com.google.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * File based implementation of secure store. Uses Java JCEKS based keystore.
+ * The data and its metadata are both stored in the keystore separately.
+ * The data is stored with the provided name as the key and the metadata is
+ * stored with the provided name + "_metadata" as the key. They are stored separately
+ * because they are accessed separately most of the time.
+ *
+ * When the client calls a put, the key is put in the keystore, the metadata object
+ * for that key is created and that is put in the store too. The system then flushes
+ *  the keystore to the file system.
+ *  During the flush, the current file is first backed up (_OLD). Then the keystore
```

Comment:
During the flush the steps are
1. Mark any existing _OLD or _NEW as orphaned, they will exist only if something had failed in the last run.
2. Move the current file to _OLD.
3. Try to write the current keystore in a _NEW file.
4. If something fails then rename the _OLD back to the main file.
5. If everything is OK then rename the _NEW to the main file and delete the _OLD file.